v2/delay/delay.go

```diff
@@ -151,39 +138,77 @@ func fileKey(file string) (string, error) {
 	return modVersionPat.ReplaceAllString(file, ""), nil
 }
 
-// Func declares a new Function. The second argument must be a function with a
-// first argument of type context.Context.
-// This function must be called at program initialization time. That means it
-// must be called in a global variable declaration or from an init function.
-// This restriction is necessary because the instance that delays a function
-// call may not be the one that executes it. Only the code executed at program
-// initialization time is guaranteed to have been run by an instance before it
-// receives a request.
+// Func declares a new function that can be called in a deferred fashion.
+// The second argument i must be a function with the first argument of
+// type context.Context.
+// To make the key globally unique, the SDK code will combine "key" with
+// the filename of the file in which myFunc is defined
+// (e.g., /some/path/myfile.go). This is convenient, but can lead to
+// failed deferred tasks if you refactor your code, or change from
+// GOPATH to go.mod, and then re-deploy with in-flight deferred tasks.
+//
+// This function Func must be called in a global scope to properly
+// register the function with the framework.
+//
+// Deprecated: Use MustRegister instead.
 func Func(key string, i interface{}) *Function {
-	f := &Function{fv: reflect.ValueOf(i)}
-
 	// Derive unique, somewhat stable key for this func.
 	_, file, _, _ := runtime.Caller(1)
 	fk, err := fileKey(file)
 	if err != nil {
 		// Not fatal, but log the error
 		stdlog.Printf("delay: %v", err)
 	}
-	f.key = fk + ":" + key
+	key = fk + ":" + key
+	f, err := registerFunction(key, i)
+	if err != nil {
+		return f
+	}
+	if old := funcs[f.key]; old != nil {
+		old.err = fmt.Errorf("multiple functions registered for %s", key)
+	}
+	funcs[f.key] = f
+	return f
+}
+
+// MustRegister declares a new function that can be called in a deferred fashion.
+// The second argument i must be a function with the first argument of
+// type context.Context.
+// MustRegister requires the key to be globally unique.
+//
+// This function MustRegister must be called in a global scope to properly
+// register the function with the framework.
+// See the package notes above for more details.
+func MustRegister(key string, i interface{}) *Function {
+	f, err := registerFunction(key, i)
+	if err != nil {
+		return f
```

Comment:
This should be a panic too?

Reply:
ugh, now that we've re-named it to MustRegister, it's really weird that it doesn't panic on these errors. The expectation of Must functions is that they will panic instead of returning errors. Is it wise for us to subvert that expectation? 🤔 

The non-panicing behavior is *slightly* less bad in these cases, because at least (unlike key collisions) 1 package can't break a function in a totally different package.  That was particularly egregious.

That said, my rationale for panicing immediately rather than proceeding to register an unusable function holds true for these errors as well.  I still cannot imagine any situation where it would be preferable for an application to deploy and start successfully when we already know the code is broken.  This is essentially just following the https://en.wikipedia.org/wiki/Fail-fast principle.

I'm happy to be convinced otherwise if someone can provide a rationale or use case where not failing fast is actually beneficial.

Reply:
Good points. I agree.

Reply:
If we decide to panic upon all error, what about making this Register and returns errors so user can do their own error handling? This way, users get more control and if we have the chance to let users handle errors gracefully with this new API, then why not.